### PR TITLE
[feature](Nereids) add cbo rewrite framework

### PR DESCRIPTION
**Pull request by morrySnow**
## Proposed changes

Issue Number: close #xxx

### the changes in this PR:
1. rename BatchRewriteJob to AbstractBatchJobExecutor
2. add a new rewrite job type, CostBasedRewriteJob. It receive a RewriteJob as input, compare the cost of two candidate plans using or not using the input RewriteJob and return the lower cost plan as the rewrite result.
3. do some small refactor on NereidsPlanner for better abstraction 
4. do some refactor on dir structure of Nereids

### the usage of cbo rewrite framework:
if you want let a rule or a rule list to be run in cbo rewrite frame work, you just need to wrap the rule / rule list with costBased function of class Rewriter, for example
```java
...
costBased(
       custom(RuleType.AGG_SCALAR_SUBQUERY_TO_WINDOW_FUNCTION,
                AggScalarSubQueryToWindowFunction::new)
),
...
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by morrySnow** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

**Review by englefly — DISMISSED** [review]

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/rewrite/CostBasedRewriteJob.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+        // Do optimize on 2 candidates
+        new Optimizer(skipCboRuleCtx).execute();
+        new Optimizer(applyCboRuleCtx).execute();
+        Optional<Pair<Cost, GroupExpression>> skipCboRuleCost = skipCboRuleCtx.getMemo().getRoot()
+                .getLowestCostPlan(PhysicalProperties.GATHER);
+        Optional<Pair<Cost, GroupExpression>> appliedCboRuleCost = applyCboRuleCtx.getMemo().getRoot()
+                .getLowestCostPlan(PhysicalProperties.GATHER);
+        // If one of them optimize failed, just return
+        if (!skipCboRuleCost.isPresent() || !appliedCboRuleCost.isPresent()) {
+            return;
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/rewrite/CostBasedRewriteJob.java:77]: right, wo should do some log here

File: fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/rewrite/CostBasedRewriteJob.java [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        new Rewriter(applyCboRuleCtx, withCboJobs).execute();
+        if (skipCboRuleCtx.getRewritePlan().deepEquals(applyCboRuleCtx.getRewritePlan())) {
+            // this means rewrite do not do anything
+            return;
+        }
+        // Do optimize on 2 candidates
+        new Optimizer(skipCboRuleCtx).execute();
+        new Optimizer(applyCboRuleCtx).execute();
+        Optional<Pair<Cost, GroupExpression>> skipCboRuleCost = skipCboRuleCtx.getMemo().getRoot()
+                .getLowestCostPlan(PhysicalProperties.GATHER);
```
[the remark this answers is not in the document]
Comment (morrySnow) [reply fe/fe-core/src/main/java/org/apache/doris/nereids/jobs/rewrite/CostBasedRewriteJob.java:72]: u r right

**Comment by morrySnow** [comment]
run buildall

**Comment by morrySnow** [comment]
run buildall

**Review by englefly — APPROVED** [review]

**Pull request merged**